File: Data Structures/Linked Lists/Singly Linked Lists/single-linked-list.go
// Double Linked Lists

package main

import (
	"fmt"
)

type Node struct {
	key  interface{}
	next *Node
}

type SingleLinkedList struct {
	head *Node
}

func (l *SingleLinkedList) Insert(key interface{}) {
	new_node := &Node{key, l.head}
	l.head = new_node
}

func (l *SingleLinkedList) Append(key interface{}) {
	new_node := &Node{key, nil}
	if l.head == nil {
		l.head = new_node
		return
	}

	last := l.head
	for last.next != nil {
		last = last.next
	}
	last.next = new_node
	return
}

func (l *SingleLinkedList) Delete(key interface{}) {
	temp := l.head
	prev := &Node{}
	if temp != nil && temp.key == key {
		l.head = temp.next
		return
	}
	for temp != nil && temp.key != key {
		prev = temp
		temp = temp.next
	}
	if temp == nil {
		return
	}
	prev.next = temp.next
}

func (l *SingleLinkedList) Show() {
	SingleLinkedList := l.head
	for SingleLinkedList != nil {
		fmt.Printf("%+v -> ", SingleLinkedList.key)
		SingleLinkedList = SingleLinkedList.next
	}
	fmt.Println()
}

func main() {
	l := SingleLinkedList{}
	l.Insert(1)
	l.Insert(2)
	l.Insert(3)
	l.Insert(4)
	l.Insert(5)
	l.Insert(6)
	l.Append(9)
	l.Delete(5)
	l.Show()
}

File: Data Structures/Linked Lists/Singly Linked Lists/single-linked-list.js
/*
    Single Linked List
    - Operations
        - Push at head of list
        - Insert after a node
        - Append node at end
        - Delete node with given key
        - Delete node at given position
*/

function SingleLinkedList() {
  var size = 0
  var head = null

  var Node = function(val) {
    this.val = val
    this.next = null
  }

  this.size = function() {
    return size
  }

  this.head = function() {
    return head
  }

  this.push = function(val) {
    var node = new Node(val)
    if(head === null){
        head = node
    } else {
        var currentNode = head

        while(currentNode.next){
            currentNode  = currentNode.next
        }

        currentNode.next = node
    }

    size++
  }

  this.remove = function(val){
    var currentNode = head
    var previousNode
    if(currentNode.val === val){
        head = currentNode.next
    } else {
      while(currentNode.val !== val) {
          previousNode = currentNode
          currentNode = currentNode.next
      }
      previousNode.next = currentNode.next
    }
    size --
  }

  this.positionOf = function(val) {
    var currentNode = head
    var index = -1

    while (currentNode) {
      index++
      if(currentNode.val === val){
          return index
      }
      currentNode = currentNode.next
    }
    return -1
  }

  this.insertAt = function(index, val){
    var node = new Node(val)

    var currentNode = head
    var previousNode
    var currentIndex = 0

    if(index > size){
        return false
    }

    if(index === 0){
        node.next = currentNode
        head = node
    } else {
        while(currentIndex < index){
            currentIndex++
            previousNode = currentNode
            currentNode = currentNode.next
        }
        node.next = currentNode
        previousNode.next = node
    }
    size++
    return true
  }

  this.removeAt = function(index) {
    var currentNode = head
    var previousNode
    var currentIndex = 0
    if (index < 0 || index >= size){
        return null
    }
    if(index === 0){
        head = currentNode.next
    } else {
        while(currentIndex < index) {
            currentIndex ++
            previousNode = currentNode
            currentNode = currentNode.next
        }
        previousNode.next = currentNode.next
    }
    size--
    return currentNode.val
  }

  this.display = function() {
    var currentNode = head
    var list = ''
    while (currentNode !== null) {
      list += currentNode.val + " => "
      currentNode = currentNode.next
    }
    console.log(list)
  }

}

var list = new SingleLinkedList()
list.push('Google')
list.push('Facebook')
list.push('Stack Overflow')
list.push('Gmail')
list.push('Github')
console.log(list.size())
console.log(list.removeAt(3))
console.log(list.positionOf('Gmail'))
console.log(list.size())
if (!list.insertAt(2, 'Added')) {
  console.log('Index Out of range!!')
}
list.display()
